src/legacy/core_plugins/kibana/public/dashboard/dashboard_app_controller.tsx

```diff
@@ -234,6 +257,15 @@ export class DashboardAppController {
         if (!isErrorEmbeddable(container)) {
           dashboardContainer = container;
 
+          dashboardContainer.renderEmpty = () => {
```

Comment:
It feels like we are mixing responsibilities here. From my perspective, the dashboard app and the dashboard container are two separate pieces of code that serve different purposes.

Is it absolutely necessary the container knows about the empty state panel? Couldn't we make that a private detail of the app to keep the interface between the container and the app as small as possible (not injecting the renderEmpty function in a kind of dirty way into the container)? For example we could not render the container at all as long as it's empty and directly render the empty screen from the app itself in line 317:
instead of
```ts
container.render(dashboardDom);
```

we are calling

```ts
if (isEmpty) {
 reactDom.render(<DashboardEmptyScreen {...propsNStuff} />, dashboardDom);
} else {
  container.render(dashboardDom);
}
```

Now if there is an important reason the dashboard container has to know it can be empty, would it make more sense have the empty screen component living in the dashboard container itself instead of the dashboard app? In that scenario the app would just pass in a boolean prop like `showEmptyScreen` or something like this and the container would take care of the rest.

Reply:
Please see one of my comments above. Empty screen needs to live inside the `dashboardContainer`, mainly for the reporting to work properly. `data-shared-items` container needs to remain where it is. Removing that wrapper will cause the report to fail. Putting it around `DashboardEmptyScreen` means the content of the empty screen will be included in the report.

Reply:
> Now if there is an important reason the dashboard container has to know it can be empty, would it make more sense have the empty screen component living in the dashboard container itself instead of the dashboard app? In that scenario the app would just pass in a boolean prop like showEmptyScreen or something like this and the container would take care of the rest.

I am not quite sure I fully understand this question, but I think there are two things to note:
1.  The props of the empty screen are still unfortunately defined on the `dashboard_app_controller`: https://github.com/elastic/kibana/pull/51939/files#diff-b2738a4dec941f426bf944b70d057b75R185
2. `DasbhoardContainer` is receiving `renderEmpty` in "a hacky way" as it can't go through input, because all inputs must be serializable